dbms/src/Interpreters/Context.cpp

```diff
@@ -1850,6 +1855,26 @@ void Context::initializeSystemLogs()
     shared->system_logs.emplace(*global_context, getConfigRef());
 }
 
+void Context::createSystemLogs()
+{
+    auto lock = getLock();
+    shared->system_logs.emplace(*global_context, getConfigRef());
```

Comment:
Hm, it is better to set `shared->system_logs` one everything initialized, since otherwise a lot of things can happen, for example deadlock which can be produced with existing data for some system tables and different structure, so that the rename should be performed, and in this case rename stacktrace will be the following:

```
Thread 1 (Thread 0x7f0c3f808040 (LWP 22383)):
6  DB::RWLockImpl::getLock () at ../dbms/src/Common/RWLock.cpp:213
7  0x0000000008dd245d in DB::IStorage::lockExclusively ()
8  0x0000000008b2cae2 in DB::InterpreterRenameQuery::execute ()
9  0x000000000501893b in DB::SystemLog<DB::TraceLogElement>::prepareTable ()
10 0x0000000004ffea92 in DB::SystemLogs::initializeSystemLogs (this=0x7f0c3f577e98, which=25)
11 0x0000000008aa0346 in DB::Context::createSystemLogs (this=0x7f0c3f578400)
12 0x0000000004f4dcdc in DB::Server::main (this=<optimized out>)
13 0x000000000bbc3013 in Poco::Util::Application::run () at ../contrib/poco/Util/src/Application.cpp:334
14 0x0000000004ffecae in DB::Server::run () at ../dbms/programs/server/Server.cpp:178
15 0x0000000004ff419c in mainEntryClickHouseServer (argc=3, argv=0x7f0c3f458bc0) at ../dbms/programs/server/Server.cpp:1060
```

So it has the following lock order:
- `Context::getLock` (in createSysmteLogs)
- `IStorage::lockExclusively`

However somewhere in parallel the following can happen:

```
Thread 16 (Thread 0x7f0c33dfb700 (LWP 22407)):
6  DB::Context::getLock (this=0x7f0c3f578400) at ../dbms/src/Interpreters/Context.cpp:466
7  0x0000000008aa28aa in DB::Context::getStoragePolicy (this=this@entry=0x7f0c3f578400, name=...) at ../dbms/src/Interpreters/Context.cpp:1707
8  0x0000000008f5fbc1 in DB::MergeTreeData::getStoragePolicy (this=0x7f0c37e31000)
9  0x0000000008f929a4 in DB::MergeTreeDataMergerMutator::getMaxSourcePartsSizeForMerge () at ../dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp:179
10 0x0000000008e61046 in DB::StorageMergeTree::merge () at ../dbms/src/Storages/StorageMergeTree.cpp:560
11 0x0000000008e68c3b in DB::StorageMergeTree::mergeMutateTask (this=0x7f0c37e31000)
... BackgroundProcessingPool ...
27 0x00007f0c3f819fb7 in start_thread (arg=<optimized out>)
28 0x00007f0c3fa711af in clone (
```

While this has the following lock order:
- `IStorage::lockExclusively` (in StorageMergeTree::merge)
- `Context::getLock`

*P.S. this is actually the real thing*

Reply:
And acquire the lock after everything initialized of course